// @unovis/ts@1.6.2/components/timeline/config.js
import { XYComponentDefaultConfig } from '../../core/xy-component/config.js';
import { TextAlign } from '../../types/text.js';
import { Arrangement } from '../../types/position.js';

const TimelineDefaultConfig = Object.assign(Object.assign({}, XYComponentDefaultConfig), { id: undefined, 
    // Items (Lines)
    cursor: undefined, type: (d) => d.type, length: (d) => d.length, color: (d) => d.color, lineRow: undefined, lineDuration: undefined, lineWidth: 8, lineCap: false, lineCursor: undefined, showEmptySegments: false, showEmptySegmentsCorrectPosition: true, lineStartIcon: undefined, lineStartIconColor: undefined, lineStartIconSize: undefined, lineStartIconArrangement: Arrangement.Inside, lineEndIcon: undefined, lineEndIconColor: undefined, lineEndIconSize: undefined, lineEndIconArrangement: Arrangement.Inside, 
    // Rows
    rowHeight: 22, alternatingRowColors: true, 
    // Row Labels
    showLabels: false, labelWidth: undefined, maxLabelWidth: 120, showRowLabels: undefined, rowLabelFormatter: undefined, rowIcon: undefined, rowLabelStyle: undefined, rowLabelWidth: undefined, rowMaxLabelWidth: undefined, rowLabelTextAlign: TextAlign.Right, 
    // Arrows
    arrows: undefined, 
    // Animation
    animationLineEnterPosition: undefined, 
    // Callbacks
    onScroll: undefined });

export { TimelineDefaultConfig };
