// @unovis/ts@1.6.2/components/graph/types.js
var GraphLayoutType;
(function (GraphLayoutType) {
    GraphLayoutType["Circular"] = "circular";
    GraphLayoutType["Concentric"] = "concentric";
    GraphLayoutType["Parallel"] = "parallel";
    GraphLayoutType["ParallelHorizontal"] = "parallel horizontal";
    GraphLayoutType["Dagre"] = "dagre";
    GraphLayoutType["Force"] = "force";
    GraphLayoutType["Elk"] = "elk";
    GraphLayoutType["Precalculated"] = "precalculated";
})(GraphLayoutType || (GraphLayoutType = {}));
var GraphLinkStyle;
(function (GraphLinkStyle) {
    GraphLinkStyle["Dashed"] = "dashed";
    GraphLinkStyle["Solid"] = "solid";
})(GraphLinkStyle || (GraphLinkStyle = {}));
var GraphLinkArrowStyle;
(function (GraphLinkArrowStyle) {
    GraphLinkArrowStyle["Single"] = "single";
    GraphLinkArrowStyle["Double"] = "double";
})(GraphLinkArrowStyle || (GraphLinkArrowStyle = {}));
var GraphNodeShape;
(function (GraphNodeShape) {
    GraphNodeShape["Circle"] = "circle";
    GraphNodeShape["Square"] = "square";
    GraphNodeShape["Hexagon"] = "hexagon";
    GraphNodeShape["Triangle"] = "triangle";
})(GraphNodeShape || (GraphNodeShape = {}));
var GraphNodeSelectionHighlightMode;
(function (GraphNodeSelectionHighlightMode) {
    GraphNodeSelectionHighlightMode["None"] = "none";
    GraphNodeSelectionHighlightMode["Greyout"] = "greyout";
    GraphNodeSelectionHighlightMode["GreyoutNonConnected"] = "greyout-non-connected";
})(GraphNodeSelectionHighlightMode || (GraphNodeSelectionHighlightMode = {}));
var GraphFitViewAlignment;
(function (GraphFitViewAlignment) {
    GraphFitViewAlignment["Center"] = "center";
    GraphFitViewAlignment["Top"] = "top";
    GraphFitViewAlignment["Bottom"] = "bottom";
    GraphFitViewAlignment["Left"] = "left";
    GraphFitViewAlignment["Right"] = "right";
})(GraphFitViewAlignment || (GraphFitViewAlignment = {}));

export { GraphFitViewAlignment, GraphLayoutType, GraphLinkArrowStyle, GraphLinkStyle, GraphNodeSelectionHighlightMode, GraphNodeShape };
