// @tanstack/vue-query-devtools@6.1.2/dist/esm/devtoolsPanel.vue.js
import { defineComponent, computed, ref, watchEffect, onMounted, onScopeDispose, createElementBlock, openBlock, normalizeStyle } from "vue";
import { useQueryClient, onlineManager } from "@tanstack/vue-query";
import { TanstackQueryDevtoolsPanel } from "@tanstack/query-devtools";
const _sfc_main = /* @__PURE__ */ defineComponent({
  __name: "devtoolsPanel",
  props: {
    client: {},
    errorTypes: {},
    styleNonce: {},
    shadowDOMTarget: {},
    style: {},
    onClose: { type: Function },
    hideDisabledQueries: { type: Boolean },
    theme: {}
  },
  setup(__props) {
    const props = __props;
    const style = computed(() => {
      return {
        height: "500px",
        ...props.style
      };
    });
    const div = ref();
    const client = props.client || useQueryClient();
    const devtools = new TanstackQueryDevtoolsPanel({
      client,
      queryFlavor: "Vue Query",
      version: "5",
      onlineManager,
      buttonPosition: "bottom-left",
      position: "bottom",
      initialIsOpen: true,
      errorTypes: props.errorTypes,
      styleNonce: props.styleNonce,
      shadowDOMTarget: props.shadowDOMTarget,
      hideDisabledQueries: props.hideDisabledQueries,
      onClose: props.onClose,
      theme: props.theme
    });
    watchEffect(() => {
      devtools.setOnClose(props.onClose ?? (() => {
      }));
      devtools.setErrorTypes(props.errorTypes || []);
      devtools.setTheme(props.theme);
    });
    onMounted(() => {
      devtools.mount(div.value);
    });
    onScopeDispose(() => {
      devtools.unmount();
    });
    return (_ctx, _cache) => {
      return openBlock(), createElementBlock("div", {
        style: normalizeStyle(style.value),
        className: "tsqd-parent-container",
        ref_key: "div",
        ref: div
      }, null, 4);
    };
  }
});
export {
  _sfc_main as default
};
