// @unovis/ts@1.6.2/components/plotband/types.js
var PlotbandLabelPosition;
(function (PlotbandLabelPosition) {
    PlotbandLabelPosition["TopLeftInside"] = "top-left-inside";
    PlotbandLabelPosition["TopLeftOutside"] = "top-left-outside";
    PlotbandLabelPosition["TopInside"] = "top-inside";
    PlotbandLabelPosition["TopOutside"] = "top-outside";
    PlotbandLabelPosition["TopRightInside"] = "top-right-inside";
    PlotbandLabelPosition["TopRightOutside"] = "top-right-outside";
    PlotbandLabelPosition["RightInside"] = "right-inside";
    PlotbandLabelPosition["RightOutside"] = "right-outside";
    PlotbandLabelPosition["BottomRightInside"] = "bottom-right-inside";
    PlotbandLabelPosition["BottomRightOutside"] = "bottom-right-outside";
    PlotbandLabelPosition["BottomInside"] = "bottom-inside";
    PlotbandLabelPosition["BottomOutside"] = "bottom-outside";
    PlotbandLabelPosition["BottomLeftInside"] = "bottom-left-inside";
    PlotbandLabelPosition["BottomLeftOutside"] = "bottom-left-outside";
    PlotbandLabelPosition["LeftInside"] = "left-inside";
    PlotbandLabelPosition["LeftOutside"] = "left-outside";
})(PlotbandLabelPosition || (PlotbandLabelPosition = {}));
var PlotbandLabelOrientation;
(function (PlotbandLabelOrientation) {
    PlotbandLabelOrientation["Horizontal"] = "horizontal";
    PlotbandLabelOrientation["Vertical"] = "vertical";
})(PlotbandLabelOrientation || (PlotbandLabelOrientation = {}));

export { PlotbandLabelOrientation, PlotbandLabelPosition };
