// @tanstack/vue-query-devtools@5.91.0/dist/cjs/devtoolsPanel.vue.cjs
"use strict";
const vue = require("vue");
const vueQuery = require("@tanstack/vue-query");
const queryDevtools = require("@tanstack/query-devtools");
const _sfc_main = /* @__PURE__ */ vue.defineComponent({
  __name: "devtoolsPanel",
  props: {
    client: {},
    errorTypes: {},
    styleNonce: {},
    shadowDOMTarget: {},
    style: {},
    onClose: { type: Function },
    hideDisabledQueries: { type: Boolean }
  },
  setup(__props) {
    const props = __props;
    const style = vue.computed(() => {
      return {
        height: "500px",
        ...props.style
      };
    });
    const div = vue.ref();
    const client = props.client || vueQuery.useQueryClient();
    const devtools = new queryDevtools.TanstackQueryDevtoolsPanel({
      client,
      queryFlavor: "Vue Query",
      version: "5",
      onlineManager: vueQuery.onlineManager,
      buttonPosition: "bottom-left",
      position: "bottom",
      initialIsOpen: true,
      errorTypes: props.errorTypes,
      styleNonce: props.styleNonce,
      shadowDOMTarget: props.shadowDOMTarget,
      hideDisabledQueries: props.hideDisabledQueries,
      onClose: props.onClose
    });
    vue.watchEffect(() => {
      devtools.setOnClose(props.onClose ?? (() => {
      }));
      devtools.setErrorTypes(props.errorTypes || []);
    });
    vue.onMounted(() => {
      devtools.mount(div.value);
    });
    vue.onScopeDispose(() => {
      devtools.unmount();
    });
    return (_ctx, _cache) => {
      return vue.openBlock(), vue.createElementBlock("div", {
        style: vue.normalizeStyle(style.value),
        className: "tsqd-parent-container",
        ref_key: "div",
        ref: div
      }, null, 4);
    };
  }
});
module.exports = _sfc_main;
