// @unovis/ts@1.6.2/components/plotline/types.js
var PlotlineLineStylePresets;
(function (PlotlineLineStylePresets) {
    PlotlineLineStylePresets["Solid"] = "solid";
    PlotlineLineStylePresets["Dash"] = "dash";
    PlotlineLineStylePresets["DashDot"] = "dashDot";
    PlotlineLineStylePresets["Dot"] = "dot";
    PlotlineLineStylePresets["LongDash"] = "longDash";
    PlotlineLineStylePresets["LongDashDot"] = "longDashDot";
    PlotlineLineStylePresets["LongDashDotDot"] = "longDashDotDot";
    PlotlineLineStylePresets["ShortDash"] = "shortDash";
    PlotlineLineStylePresets["ShortDashDot"] = "shortDashDot";
    PlotlineLineStylePresets["ShortDashDotDot"] = "shortDashDotDot";
    PlotlineLineStylePresets["ShortDot"] = "shortDot";
})(PlotlineLineStylePresets || (PlotlineLineStylePresets = {}));
var PlotlineLabelPosition;
(function (PlotlineLabelPosition) {
    PlotlineLabelPosition["TopLeft"] = "top-left";
    PlotlineLabelPosition["Top"] = "top";
    PlotlineLabelPosition["TopRight"] = "top-right";
    PlotlineLabelPosition["Right"] = "right";
    PlotlineLabelPosition["BottomRight"] = "bottom-right";
    PlotlineLabelPosition["Bottom"] = "bottom";
    PlotlineLabelPosition["BottomLeft"] = "bottom-left";
    PlotlineLabelPosition["Left"] = "left";
})(PlotlineLabelPosition || (PlotlineLabelPosition = {}));
var PlotlineLabelOrientation;
(function (PlotlineLabelOrientation) {
    PlotlineLabelOrientation["Horizontal"] = "horizontal";
    PlotlineLabelOrientation["Vertical"] = "vertical";
})(PlotlineLabelOrientation || (PlotlineLabelOrientation = {}));

export { PlotlineLabelOrientation, PlotlineLabelPosition, PlotlineLineStylePresets };
